### ISSUE TITLE
A read-only audit against the vision: the remediation loop cannot close, and the human cannot answer

### ISSUE THREAD

**Issue by timwukp** [timwukp/llmops-agentic-system#60]
## Why this issue

A read-only audit of `main` (120 files, 200 commits), measured against a stated product vision rather than against the repo's own claims. Nothing was changed; every defect below was reproduced from the tree at `3a559eb`.

The repo is unusually rigorous — no action-level IAM wildcards, real dry-run diffs, `conftest.py` that makes the socket refuse to open, and `contracts/tasks.py`, which is the best "declared-but-unreachable capability" guard I have seen in a public repo. The findings below are not about care. They are about **where the care has been pointed**.

**繁體中文摘要**：這是一次唯讀盤點，對照的是產品願景而非 repo 自身的宣稱。共九項發現：四項是「已宣稱但做不到」的缺陷（自我修復迴圈必死、人類無法回話、七個 harness 全走公網、console 所有 GET 公開），五項是願景隱含但不存在的能力（logit/on-policy 蒸餾、資料融合階段、model registry 契約、教師授權閘門、分層評測）。最後有四個開放性問題需要決定，其中最重要的是：`human task driven` 指的是治理，還是抽象層？

---

## The vision this is measured against

> A human-task-driven, end-to-end agentic LLMOps platform for enterprises automating **large-model distillation** — both **open-weight teachers** (DeepSeek, Kimi, Qwen — weights visible and changeable) and **closed API teachers** (Claude, GPT). After distillation, the enterprise fine-tunes a **small open-weight student** (QLoRA on Qwen/DeepSeek/Kimi) using **its own private datasets**, and the output lands in a **model registry**, production-ready after test and evaluation.
>
> The point is to remove the enterprise's hardest and most repetitive labour: distilling the big model, and upgrading the small one with the capability learned from it plus the enterprise's own private data.

Two things follow immediately, and they reorder the priorities below:

1. **"Human-task-driven" is the differentiator, not the modifier.** The consult → priced plan → KMS-signed approval → `service_launch_run` spine is already built, and near-nothing else in this product category has it. Distillation itself is commoditising.
2. **"Weights visible and changeable" is a technical claim** about logit access. See gap 2.1.

---

### 1.1 · P0 · The remediation loop cannot reach a terminal state

`IncrementIteration` is a `Pass` with `Parameters` and **no `ResultPath`**, so its constructed object replaces the entire state — dropping `task_id` and `pipeline_mode`:

```json
"IncrementIteration": {
  "Type": "Pass",
  "Parameters": {
    "run_id.$": "$.run_id",
    "manifest_uri.$": "$.manifest_uri",
    "iteration.$": "States.MathAdd($.iteration, 1)"
  },
  "Next": "RemediateFinetune"
}
```

Three downstream terminal states read `$.task_id`: `Complete`, `MarkTaskDone` (`Key: {"id": {"S.$": "$.task_id"}}`), and `MarkTaskFailed`.

`orchestration/start_pipeline/handler.py:47-49` already documents why this is fatal — the `NO_TASK = "none"` sentinel exists precisely because *a JSONPath that is not present raises `States.Runtime`, which no `Catch` can intercept.*

**So any run that enters remediation even once dies uncatchably at closeout, leaving the task record unclosed.** This is the self-healing loop the README leads with. It is a regression introduced with `task_id` in v1.2.0 and never exercised, because no run has entered remediation since.

The only guard (`tests/test_orchestration.py:1637-1638`) asserts the `Next` transitions and nothing about field preservation.

**Proposed fix:** carry the fields through, plus a guard that walks every path through the ASL, collects every `$.field` each state reads, and asserts presence at that point. That guard generalises past this one bug, which is the house style.

### 1.2 · P0 · `escalate_human` is one-way — the human cannot answer

Every specialist prompt says *"The pipeline pauses."* It does not. `harness_driver/handler.py:650-651` calls `send_task_failure(error="EscalatedToHuman")` → `States.ALL` catch → `EscalateFail` → `MarkRunFailed` → `MarkTaskFailed` → `Fail`. **The run is over.**

There is no `Wait` state and no human `waitForTaskToken` gate anywhere in the ASL. The directive mailbox (`put_directive`/`take_directive`) is readable only by a *live* driver invocation at a voluntary `checkpoint`, and `run_can_hear_a_directive` returns `undeliverable` once the run is `escalated`/`failed`/`completed`. The `llmops-escalations` SNS topic has no subscriber unless `--escalation-email` was passed.

For a platform whose thesis is *human-task-driven*, a human can **initiate** but not **reply**. This is a hole in the thesis, not an operational nit.

**Open decision:** implement a real `HumanGate` (`waitForTaskToken` + a console resume route), or rewrite seven prompts to say the run ends. Today it is neither.

### 1.3 · P0 · Enterprise private data cannot legally flow through this

All seven harnesses are `"networkMode": "PUBLIC"`. No agent has a `harness.prod.json`, yet `05_harnesses.py --prod` looks for one and `02_network.py` provisions a VPC (~$2.64/day in interface endpoints) that nothing uses.

Credit where due: `customer-data/` handling is better than expected — `harness_execution_role.json:171-176` grants read-only with an excellent rationale ("a pipeline that can rewrite the customer's data can destroy the held-out set its own gates are judged on"), and `data-prep/audit` does a read-only Macie cross-check. The prefix discipline is right; the network path is not.

Under the vision this moves from tidiness to **blocker**: the first question in any enterprise data-classification review is the data path, and the answer today is "the public internet."

### 1.4 · P0 · Every console GET is unauthenticated

`deploy/console/lambda_function.py:11` states the design: *GET routes are public read-only.* In practice, anyone with the API id can read `/api/run`, `/api/cost-overview`, `/api/cost-estimates`, `/api/tasks`, `/api/tasks/{id}`, and `/api/tasks/{id}/approval` — run detail, spend, **customer consult threads, and signed approval records**.

PRs #52 and #53 responded to this by removing the live console URL from the repo — i.e. treating URL secrecy as the access control. Related: `BUDGET_MODE` defaults to `advisory` (over-budget plans launch anyway), and `deploy.sh` puts the single default `admin` in both `llmops-approver` and `llmops-datascience`, so four-eyes is impossible out of the box.

---

### 2.1 · The vision requires a second distillation mode that does not exist

`data-prep/generate` calls `bedrock-runtime converse` against a teacher, strips `<think>`, and SFTs the student. That is **off-policy (offline) black-box distillation**. Searching the harness prompts and pipeline for `logprob`, KL divergence, or on-policy anything returns nothing.

This matters for the vision's credibility. "Distil open-weight models, weights visible and changeable" reads, to an ML team, as **logit access** — which unlocks token-level KL, on-policy distillation, and draft-model training, none of which Bedrock Model Distillation can offer and none of which the current architecture can express.

It also reframes Phase 4's **0/16**. That was diagnosed as too few traces (6). It is *also* textbook off-policy distribution mismatch: the recent literature is consistent that the dominant off-policy paradigm trains students on static teacher output so they [never encounter their own errors during learning](https://arxiv.org/html/2604.00626v1), that on-policy distillation lets the student sample its own trajectories under dense token-level teacher supervision to [close the train/inference distribution gap](https://arxiv.org/html/2601.18734v3), and that even precomputing teacher log-probabilities offline [fails to reliably match standard OPD](http://arxiv.org/abs/2604.13010v1). *(Sources paraphrased; content was rephrased for compliance with licensing restrictions.)*

**`pipeline/v2` is a better offline dataset — still off-policy.** Worth wiring for the fastest passing model, but it is not the capability the vision promises.

Missing path: mirror open weights (`data-prep/mirror_model`, currently an ACCEPTED GAP never invoked) → self-host the teacher → extract logits → KL / on-policy distillation.

### 2.2 · `pipeline/v2` is not connected to the platform at all

Zero references into `pipeline/v2` from `agents/`, `orchestration/`, `deploy/`, or `pipeline/contracts/`. No harness prompt mentions it. `start_pipeline.DEFAULT_PARAMS` still hardcodes `dataset: arc-agi-2`, `sample_count: 2000`, and `relative_solve_rate`/`format_validity` gates.

`pipeline/v2/README.md` says "the platform side needs no new engineering: data scale is a manifest parameter." That sentence is currently false. Making it true is wiring, not research — and it is the highest-value unlock in the repo, because the one item the case study lists as *Not proved* already has its solution sitting in the same tree.

### 2.3 · No stage represents "teacher capability + private data"

The stated goal is the **combination**. The ASL goes `generate → curate → launch`; `curate` dedupes, scrubs PII, self-scores, and splits 90/10. **Nothing mixes two sources.**

The real decisions live here: blend ratio, curriculum order, catastrophic forgetting, what gives when private and teacher distributions conflict, and **contamination checks** (does teacher-generated data leak the held-out private eval set?). A `data-prep/blend` task would make these explicit manifest parameters — and blend ratio is exactly the kind of decision a human should be signing, which strengthens 1.2.

### 2.4 · The model registry is the fourth instance of this repo's signature defect

`contracts/tasks.py` names the pattern in its own docstring: *a capability the prompts, the docs and the IAM all describe, that no code path can ever reach*, with three precedents (`page_human` #54, eval `evaluate` #57, the whole monitor harness #58).

The registry is number four. IAM grants `sagemaker:CreateModelPackage` / `CreateModelPackageGroup` (`harness_execution_role.json:259-262`) and the deploy prompt says `create-model-package` — but `contracts/` has nothing, and `manifest.schema.json` top-level is `['run_id','created_at','trigger_source','plan','iteration','stages','models']`. By this repo's own standard, no contract means it does not exist yet.

The registry is a **named output** of the vision, and for an enterprise it *is* the compliance artifact. Cheap unlock: **reuse the existing KMS ECDSA + hash-chain machinery to sign model packages**, binding teacher / dataset / hyperparams / eval / licence into verifiable lineage. The signing infrastructure is already built for approvals.

### 2.5 · No teacher licence / ToS gate

The vision explicitly includes distilling Claude and GPT. Many providers' terms restrict using outputs to train competing models. Enterprise legal will ask, and an unanswered version of this question stops procurement.

Proposal: a `teacher_licence` contract captured at consult time (teacher model, licence basis, permitted use, distribution limits on the resulting weights), carried in the manifest, attached to the model package, and **included in `SIGNED_KEYS`** — so the human signs a legal position, not just a budget. Low cost, high differentiation, and precisely the repetitive labour the vision targets.

### 2.6 · Executable scoring has a ceiling most enterprise domains sit outside

`pipeline/v2/eval_student.py` replacing LLM judges with executable verification (oracle self-test 1000/1000) is the right instinct for enterprise trust. But it only works where executable ground truth exists — code, math, structured extraction. Enterprise reality is largely summarisation, support conversations, RAG answers, and fuzzy-label classification.

Left unaddressed, "we don't use LLM judges" becomes the constraint that keeps this locked to ARC-AGI-2. Suggested tiering, with each gate declaring its tier: **T1 executable → T2 rubric + calibrated judge (publishing judge/human agreement) → T3 human sampling.**

---

## Part 3 — Structural findings

- **No declarative IaC anywhere** (no `.tf`, CFN, CDK, SAM). Eight imperative scripts run by hand in a non-obvious order — `04_wire_memory.py` must run *after* `05_harnesses.py`, a fact that lives only in a README table.
- **Confirmed absent:** `pyproject.toml`, `requirements.txt`, `Makefile`, `CONTRIBUTING.md`, `deploy/deploy_all.py`, `deploy/99_teardown.py`.
- **One deployment per account+region.** Resource names are hardcoded module constants with no `--name-prefix`. Running `01_iam.py` for a second region **overwrites the first region's global IAM role policies**.
- **Teardown is prose only.** The single programmatic destroy, `02_network.py --destroy`, calls `find_by_tag` without a VPC filter — it deletes every same-tagged interface endpoint in the region.
- **`SPANS_SINCE = "2026-07-28T12:00:00Z"`** is hardcoded in `deploy.sh` and the console handler. It is the author's own OTEL-enablement date; for anyone else the observability and evaluations tabs filter against a meaningless timestamp.
- **`MODEL_FALLBACKS` failover mutates the shared harness definition globally and never reverts it**, affecting every concurrent and future run.
- **`DriftDetected` has no consumer.** Defined at `contracts/events.py:32`, emitted at `harness_driver/handler.py:397`, and no rule routes it. "Production-ready deployment" needs drift → rollback to close.
- **Two stale excuses in `contracts/tasks.py`:** `("eval","evaluate")` and `("orchestrator","triage")` both say "exists on a branch, not on main," but `EvalGenerate` is at ASL line 197 and `ensure_triage_rule` is wired at `07_lambdas.py:511`. The file is designed so entries must be *removed* to go green; these two are now overdue.

---

## Part 4 — Where the effort has been going

| Measure | Value |
|---|---|
| Test code | 21,152 lines |
| System code (`agents` + `orchestration` + `pipeline`) | ~7,900 lines |
| Ratio | **2.7 : 1** |
| `CHANGELOG.md` | 1,033 lines, **779 (75%) still `[Unreleased]`** |
| Last 10 merged PRs that are docs / guards / meta | **10 / 10** |
| Plain issues ever opened (before this one) | **0** — all 59 numbers are PRs |

The last ten PRs reworded the README, deleted a duplicated mp4, trimmed a section, removed the console URL twice, and added a guard verifying that a previous guard checked anything. That last one is admirable self-scrutiny. It is also the point where marginal return went to zero — while the headline remediation loop carried an uncatchable `States.Runtime` on `main` the whole time.

The lesson is not "fewer guards." It is that **guards only check the claims someone thought to check**, and "what happens when a run actually remediates" was never on the list.

---

## Part 5 — Open questions

**Q1 — Which reading of "human-task-driven" is the product?**
(a) A human initiates and signs every task (**governance**), or (b) the unit of work is a business outcome rather than an ML job (**abstraction**)? `main` does (a) well and has one prototype of (b) in `consult`. Both are worth building; they order the P0 list differently.

**Q2 — Is logit-level / on-policy distillation (2.1) the north star, or out of scope?**
It is the only part genuinely hard to copy, and Bedrock Model Distillation structurally cannot offer it. It also requires self-hosted teacher GPUs, which changes the cost model the FinOps agent is built around. Committing or explicitly declining changes the next two quarters.

**Q3 — What is the first target domain?**
Every gate, dataset, and prompt is shaped around ARC-AGI-2 (executable reasoning). Enterprise buyers usually arrive with support, documents, or extraction. Changing domain pulls in the tiered evaluation of 2.6.

**Q4 — Should the positioning converge?**
The seven-language narrated intro and bilingual docs optimise for a technical reviewer. Parts 1–2 optimise for an enterprise operator. Both compete for the same PRs, and lately the reviewer has won 10 out of 10. Worth an explicit split — e.g. a cap on the share of any release that is docs/guard work — and a positioning line at the top of the README either way.

---

## Suggested sequencing

| Priority | Items |
|---|---|
| **P0 — make the thesis true** | 1.1 remediation fix + path-walking guard · 1.2 `HumanGate` · 1.3 VPC harness variants · 1.4 console GET auth |
| **P1 — deliver what the vision names** | 2.2 wire `pipeline/v2` · 2.3 `data-prep/blend` · 2.4 registry contract + KMS-signed lineage · 2.5 `teacher_licence` gate · 2.6 tiered evaluation |
| **P2 — the moat** | 2.1 self-hosted teacher → logits → on-policy distillation · drift → rollback |
| **P3 — platform** | `--name-prefix` · `deploy_all.py` / `99_teardown.py` · packaging · multi-tenant · cross-run comparison (`orchestrator/report`) · extract the `AGENTS.md` gotchas bank as a standalone artifact |

---

## Method note

Read-only analysis of `3a559eb`. No files were modified and no branch was pushed. Every Part 1 defect and every "absent"/"zero references" claim was reproduced directly against the tree; a few Part 3 items (`MODEL_FALLBACKS` scope, `BUDGET_MODE` default, console route inventory) come from a full read of the deploy and console layers rather than an executed reproduction. The two external claims in 2.1 are cited inline and paraphrased.

One aside worth its own issue eventually: the gotchas bank in `AGENTS.md` — dual ARN handling, `additionalParams` silently replacing the system prompt, a bad skill source failing at session start rather than at `UpdateHarness`, `allowedTools` not accepting globs — is probably the single most reusable asset here, and it is currently a section of one file.
